### ISSUE TITLE
new provisioner with an already existing name leads to unexpected behavior

### ISSUE THREAD

**Issue by DonOtuseGH** [smallstep/certificates#886]
### Subject of the issue
It is possible to create an additional provisioner with an already existing name using `step` cli which leads to unexpected behavior.

### Your environment
* latest Docker image 
```
smallstep/step-ca    latest       1484a1b3c0c2   4 weeks ago     113MB
```
* Version
```
bash-5.1$ step -v
Smallstep CLI/0.18.2 (linux/amd64)
Release Date: 2022-03-01 23:22 UTC
```

### Steps to reproduce

create the first provisioner, reload step and list available provisioners
```
bash-5.1$ step ca provisioner list
[]

bash-5.1$ step ca provisioner add ca@example.com --type jwk --create
Please enter a password to encrypt the provisioner private key? [leave empty and we'll generate one]:
✔ Password: 'jzHymd[|.=Qe$!9'P;!P-biUY}M~U)9
Success! Your `step-ca` config has been updated. To pick up the new configuration SIGHUP (kill -1 <pid>) or restart the step-ca process.

bash-5.1$ kill -1 1
bash-5.1$ step ca provisioner list
[
   {
      "type": "JWK",
      "name": "ca@example.com",
      "key": {
         "use": "sig",
         "kty": "EC",
         "kid": "tTvVir_lNTXw0BlFueOab1i29d5EF1fWcGMYWOJYBL0",
         "crv": "P-256",
         "alg": "ES256",
         "x": "RAJbUrEhlcG-cLdBrcQf0M1w_iSHZWzMO6jO8dP4ZWU",
         "y": "krh89GSjajrr9CL-NVDBYAtcItzYDqrT2JICP9-zrcc"
      },
      "encryptedKey": "eyJhbGciOiJQQkVTMi1IUzI1NitBMTI4S1ciLCJjdHkiOiJqd2sranNvbiIsImVuYyI6IkEyNTZHQ00iLCJwMmMiOjEwMDAwMCwicDJzIjoiN0RRX3RGcWtBeDhmZWR0OUQtQ1RWZyJ9.X5TA5yQOdLFewl94ThfnQgWNwKTxy6aqUZRpPac8sI0svoP1gJv99A.qvijRHMDHJTbQOyz.8sq7C0wCfhOpB92kpTitS-61Ca1qRl3v2Er5J2NtRIU9t0o67WQIrn2xe2iOX8epQXgnPVBevlT6Bz4CmmaioJpTKTc0c2NsrBF1Rb3olMrkfTIQDexLNuTa2_gjt7-DdhhdKuU5KphF-mk3cLiEkLkoDhwMCUQs6mLtRkoKKKOVQRocXKY_Wg9u7xNfIvY807Bo0jLzMB-tYvLwfctszaJ5aQWpdhl2TRRl4bY1kWX7oaF7s2kLaCuejDntuz_zIO9orp79Pp65XrIK4ilFGIfKKWG2F8oWSZDuquE3n1RQe0TUHkbv_BNKkXHTp0pplJx39BlNcTwIV-bkg7w.OnCDEItYYQUZ-fDitIv-8g"
   }
]
```

create a certificate using the provisioner from above
```
bash-5.1$ step ca certificate www.example.com www.crt www.key
✔ Provisioner: ca@example.com (JWK) [kid: tTvVir_lNTXw0BlFueOab1i29d5EF1fWcGMYWOJYBL0]
Please enter the password to decrypt the provisioner key:
✔ CA: https://localhost:9000
✔ Certificate: www.crt
✔ Private Key: www.key
```

try to create another provisioner with an already existing name, reload step and list available provisioners
```
bash-5.1$ step ca provisioner add ca@example.com --type jwk --create
Please enter a password to encrypt the provisioner private key? [leave empty and we'll generate one]:
✔ Password: $uFM,s5jQBp*oUU#@w10M(EywY!k9m-Y
Success! Your `step-ca` config has been updated. To pick up the new configuration SIGHUP (kill -1 <pid>) or restart the step-ca process.

bash-5.1$ kill -1 1
bash-5.1$ step ca provisioner list
[
   {
      "type": "JWK",
      "name": "ca@example.com",
      "key": {
         "use": "sig",
         "kty": "EC",
         "kid": "tTvVir_lNTXw0BlFueOab1i29d5EF1fWcGMYWOJYBL0",
         "crv": "P-256",
         "alg": "ES256",
         "x": "RAJbUrEhlcG-cLdBrcQf0M1w_iSHZWzMO6jO8dP4ZWU",
         "y": "krh89GSjajrr9CL-NVDBYAtcItzYDqrT2JICP9-zrcc"
      },
      "encryptedKey": "eyJhbGciOiJQQkVTMi1IUzI1NitBMTI4S1ciLCJjdHkiOiJqd2sranNvbiIsImVuYyI6IkEyNTZHQ00iLCJwMmMiOjEwMDAwMCwicDJzIjoiN0RRX3RGcWtBeDhmZWR0OUQtQ1RWZyJ9.X5TA5yQOdLFewl94ThfnQgWNwKTxy6aqUZRpPac8sI0svoP1gJv99A.qvijRHMDHJTbQOyz.8sq7C0wCfhOpB92kpTitS-61Ca1qRl3v2Er5J2NtRIU9t0o67WQIrn2xe2iOX8epQXgnPVBevlT6Bz4CmmaioJpTKTc0c2NsrBF1Rb3olMrkfTIQDexLNuTa2_gjt7-DdhhdKuU5KphF-mk3cLiEkLkoDhwMCUQs6mLtRkoKKKOVQRocXKY_Wg9u7xNfIvY807Bo0jLzMB-tYvLwfctszaJ5aQWpdhl2TRRl4bY1kWX7oaF7s2kLaCuejDntuz_zIO9orp79Pp65XrIK4ilFGIfKKWG2F8oWSZDuquE3n1RQe0TUHkbv_BNKkXHTp0pplJx39BlNcTwIV-bkg7w.OnCDEItYYQUZ-fDitIv-8g"
   }
]
```

re-create the certificate with the password from 2nd provisioner fails, but succeeds using the password from the 1st/initial provisioner
```
bash-5.1$ step ca certificate www.example.com www.crt www.key --force
✔ Provisioner: ca@example.com (JWK) [kid: tTvVir_lNTXw0BlFueOab1i29d5EF1fWcGMYWOJYBL0]
Please enter the password to decrypt the provisioner key: $uFM,s5jQBp*oUU#@w10M(EywY!k9m-Y
Please enter the password to decrypt the provisioner key: 'jzHymd[|.=Qe$!9'P;!P-biUY}M~U)9
✔ CA: https://localhost:9000
✔ Certificate: www.crt
✔ Private Key: www.key

```
 
restart of Docker container fails, because the 2nd provisioner has been saved to `ca.json`
```
$ docker-compose restart && docker-compose logs -f
...
step-ca    | 2022/04/01 11:30:34 shutting down ...
step-ca    | badger 2022/04/01 11:30:34 INFO: Storing value log head: {Fid:0 Len:31 Offset:30253}
step-ca    | badger 2022/04/01 11:30:34 INFO: [Compactor: 173] Running compaction: {level:0 score:1.73 dropPrefixes:[]} for level: 0
step-ca    | badger 2022/04/01 11:30:34 INFO: LOG Compact 0->1, del 2 tables, add 1 tables, took 2.287264ms
step-ca    | badger 2022/04/01 11:30:34 INFO: [Compactor: 173] Compaction for level: 0 DONE
step-ca    | badger 2022/04/01 11:30:34 INFO: Force compaction on level 0 done
step-ca    | badger 2022/04/01 11:30:35 INFO: All 1 tables opened in 1ms
step-ca    | badger 2022/04/01 11:30:35 INFO: Replaying file id: 0 at offset: 30284
step-ca    | badger 2022/04/01 11:30:35 INFO: Replay took: 8.091��s
step-ca    | cannot add multiple provisioners with the same name
step-ca    | badger 2022/04/01 11:30:36 INFO: All 1 tables opened in 1ms
step-ca    | badger 2022/04/01 11:30:36 INFO: Replaying file id: 0 at offset: 30284
step-ca    | badger 2022/04/01 11:30:36 INFO: Replay took: 58.315��s
step-ca    | cannot add multiple provisioners with the same name
step-ca    | badger 2022/04/01 11:30:36 INFO: All 1 tables opened in 1ms
step-ca    | badger 2022/04/01 11:30:36 INFO: Replaying file id: 0 at offset: 30284
step-ca    | badger 2022/04/01 11:30:36 INFO: Replay took: 71.906��s
step-ca    | cannot add multiple provisioners with the same name
step-ca    | badger 2022/04/01 11:30:37 INFO: All 1 tables opened in 1ms
step-ca    | badger 2022/04/01 11:30:37 INFO: Replaying file id: 0 at offset: 30284
step-ca    | badger 2022/04/01 11:30:37 INFO: Replay took: 107.468��s
step-ca    | cannot add multiple provisioners with the same name
step-ca    | badger 2022/04/01 11:30:38 INFO: All 1 tables opened in 1ms
step-ca    | badger 2022/04/01 11:30:38 INFO: Replaying file id: 0 at offset: 30284
step-ca    | badger 2022/04/01 11:30:38 INFO: Replay took: 109.141��s
step-ca    | cannot add multiple provisioners with the same name
step-ca    | badger 2022/04/01 11:30:41 INFO: All 1 tables opened in 0s
step-ca    | badger 2022/04/01 11:30:41 INFO: Replaying file id: 0 at offset: 30284
step-ca    | badger 2022/04/01 11:30:41 INFO: Replay took: 160.021��s
step-ca    | cannot add multiple provisioners with the same name

$ cat ca.json
...
	"authority": {
		"provisioners": [
			{
				"type": "JWK",
				"name": "ca@example.com",
				"key": {
					"use": "sig",
					"kty": "EC",
					"kid": "tTvVir_lNTXw0BlFueOab1i29d5EF1fWcGMYWOJYBL0",
					"crv": "P-256",
					"alg": "ES256",
					"x": "RAJbUrEhlcG-cLdBrcQf0M1w_iSHZWzMO6jO8dP4ZWU",
					"y": "krh89GSjajrr9CL-NVDBYAtcItzYDqrT2JICP9-zrcc"
				},
				"encryptedKey": "eyJhbGciOiJQQkVTMi1IUzI1NitBMTI4S1ciLCJjdHkiOiJqd2sranNvbiIsImVuYyI6IkEyNTZHQ00iLCJwMmMiOjEwMDAwMCwicDJzIjoiN0RRX3RGcWtBeDhmZWR0OUQtQ1RWZyJ9.X5TA5yQOdLFewl94ThfnQgWNwKTxy6aqUZRpPac8sI0svoP1gJv99A.qvijRHMDHJTbQOyz.8sq7C0wCfhOpB92kpTitS-61Ca1qRl3v2Er5J2NtRIU9t0o67WQIrn2xe2iOX8epQXgnPVBevlT6Bz4CmmaioJpTKTc0c2NsrBF1Rb3olMrkfTIQDexLNuTa2_gjt7-DdhhdKuU5KphF-mk3cLiEkLkoDhwMCUQs6mLtRkoKKKOVQRocXKY_Wg9u7xNfIvY807Bo0jLzMB-tYvLwfctszaJ5aQWpdhl2TRRl4bY1kWX7oaF7s2kLaCuejDntuz_zIO9orp79Pp65XrIK4ilFGIfKKWG2F8oWSZDuquE3n1RQe0TUHkbv_BNKkXHTp0pplJx39BlNcTwIV-bkg7w.OnCDEItYYQUZ-fDitIv-8g"
			},
			{
				"type": "JWK",
				"name": "ca@example.com",
				"key": {
					"use": "sig",
					"kty": "EC",
					"kid": "tDAoZ5cl3QFoPz6O0D_iMT9DHXnzQ_zpmPSnbSsJ21E",
					"crv": "P-256",
					"alg": "ES256",
					"x": "k7cCOcaoicgEnnTQDY4FBeyisx08FeVHYcFtdetKkcs",
					"y": "Ar9BJjyfIEECus5tXEuZukAjj-F9xbIVTRgsQb7AlD0"
				},
				"encryptedKey": "eyJhbGciOiJQQkVTMi1IUzI1NitBMTI4S1ciLCJjdHkiOiJqd2sranNvbiIsImVuYyI6IkEyNTZHQ00iLCJwMmMiOjEwMDAwMCwicDJzIjoiN0xieE43eGlRbkdNOGJqS1BBeDlmUSJ9.CFnFwY2h7Kbl_dfuPENCJDFwLLr-yVbBVgSxA3ZqeTnJU6IyXNN0zQ.NQzR851rzSKEkWjq.i4EC1c50Hp4gw7M_tur7c1qVrCQ7QC0RbS2_7cPjN1EyiyRVTOalhBOUpqMSza6Woy9_6OvjSQ9Qad7SSmx5RaRUFhc9c_n2qD3LbWlX7uPe-dymLdt1MxabgsAS0B0Vfuwzbi6DNE9xmZdYbuVNlCNo9QxKMiHgY3UxuLTXNKk6TrtkwMZWCF3oPReHwJV2OeZaN1grcLBufjcb6GYR6x1DDWeTCL9iYLT8hdAYDBrQICL0M2XIiBjqjz6xPKTJDdngeAdM1aqnbe772IbEWaSuyxi0frZRuWVuLz2dyRcXTWEYLA0JGIzK188pJe_0OHgn-RyPT6qpOSILMQo.GTI9IhWPIMho9uyGN6zVSQ"
			}
		],
```

### Expected behaviour
`step` cli has to prevent the creation of provisioners with an already existing name and display a corresponding error message

### Actual behaviour
`step`cli allows creation of provisioners with an already existing name, does not show any error message or list them, but saves the new provisioner in `ca.json`, which prevents the container/service from being restarted properly.

**Comment by dopey** [comment]
Hey @DonOtuseGH thanks for opening the issue. There are a few things at play here causing the "unexpected behavior".

1. Historically, all CA configuration has been stored in the `ca.json`. Any misconfiguration (using `step ca provisioner [add|remove] could be fixed by simply editing the ca.json, so it's impossible to get into an irrecoverable state.
2. Recently we added the ability to manage the CA via an API and store the configuration in a DB layer. Now there are configurations that cannot be recovered from without directly making changes in the DB.
3. We added a new subset of provisioner commands to `step beta ca provisioner ...` to be used with the new database managed configuration. 
4. The original command subcommands still exist (`step ca provisioner ...`) but only work on the `ca.json` configuration.
5. We didn't go and backport all the logical consistency checks to the `step ca provisioner` command group because users can always open the `ca.json` to fix an issue.

tl;dr You're right. We should be catching this instead of adding the provisioner then failing to start. The current workaround is to pop open your ca.json and rename the new provisioner to a unique name. I'm currently working on synchronizing the original and the `beta` command groups, so this probably should be resolved soon.

**Comment by DonOtuseGH** [comment]
@dopey Thanks for your answer. Of course, it wasn't a big deal to fix it in the `ca.json` file. I simply wanted to point out the problem and thus offer the possibility to fix it. 😁 I also found the `step beta` stuff and i like the possibility to configure provisioner claims. Additionally, it would be very handy to configure other options like templates for the provisioners as well. 👍
